File: Roadmap/02 - FUNCIONES Y ALCANCE/java/AndrewCodev.java
public class AndrewCodev {
	//Variables Globales de la clase
	int num1;
    int num2;
    int num3;

    public void suma(){
    	this.num1 = 2;
    	this.num2 = 7;
    	this.num3 = 1;
    	
    	int resultado = this.num1 + this.num2 + this.num3;
        System.out.println("Resultado de la suma en un método sin argumentos y sin retorno");
        System.out.println(this.num1+" + "+this.num2+" + "+this.num3+" = "+ resultado);
    }
    
    /*las variables creadas como parametros son variable locales del método
     * al igual que las variables creadas dentro de este sin el operador this.
     * como ejemplola variable resultado.
     */
    public void sumaConParametros(int num1, int num2, int num3) {
    	//Variable Local llamando a la función Suma con Retorno
    	int resultado = sumaConRetorno(num1, num2, num3);
        System.out.println(this.num1+" + "+this.num2+" + "+this.num3+" = "+ resultado);
    }
    
    public int sumaConRetorno(int num1, int num2, int num3) {
    	this.num1 = num1;
    	this.num2 = num2;
    	this.num3 = num3;
    	int resultado = this.num1 + this.num2 + this.num3;
        return resultado;
    }
    
    //DIFICULTAD EXTRA
    public int contadorNumerico(String cadena1, String cadena2) {
    	//Variable local de método
    	int contador = 0;
    	
    	for (int i = 1; i <= 100; i++) {
    		
    		if(i % 3 == 0 && i % 2 == 0) {
    			System.err.println(i+": "+cadena1+" "+cadena2);
    		}else if(i % 2 == 0) {
    			System.err.println(i+": "+cadena2);
    		}else if(i % 3 == 0) {
    			System.err.println(i+": "+cadena1);
    		}
    		else {
    			System.err.println(i+" ");
    			contador++;
    		}			
		}
    	
    	return contador;
    }
    
    public static void main(String[] args) {
		AndrewCodev andrewCodev = new AndrewCodev();
		andrewCodev.suma();
		System.out.println("\nResultado de la suma en un método con argumentos y sin retorno");
		andrewCodev.sumaConParametros(5, 5, 10);
		

    	System.out.println("\nResultado de la suma en un método con argumentos y con retorno\n");
		int sumaConRetorno = andrewCodev.sumaConRetorno(6, 4, 5);
		System.out.println(sumaConRetorno);
		
		System.out.println("DIFICULTAD EXTRA");
		int contador = andrewCodev.contadorNumerico("Hola Mundo", "Adios Mundo");
		System.err.println("Total numeros no multiplos de 2 ni de 3: "+contador);
    }

}

File: Roadmap/02 - FUNCIONES Y ALCANCE/javascript/AndrewCodev.js
/*
 * EJERCICIO:
 * - Crea ejemplos de funciones básicas que representen las diferentes
 *   posibilidades del lenguaje:
 *   Sin parámetros ni retorno, con uno o varios parámetros, con retorno...
 * - Comprueba si puedes crear funciones dentro de funciones.
 * - Utiliza algún ejemplo de funciones ya creadas en el lenguaje.
 * - Pon a prueba el concepto de variable LOCAL y GLOBAL.
 * - Debes hacer print por consola del resultado de todos los ejemplos.
 *   (y tener en cuenta que cada lenguaje puede poseer más o menos posibilidades)
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea una función que reciba dos parámetros de tipo cadena de texto y retorne un número.
 * - La función imprime todos los números del 1 al 100. Teniendo en cuenta que:
 *   - Si el número es múltiplo de 3, muestra la cadena de texto del primer parámetro.
 *   - Si el número es múltiplo de 5, muestra la cadena de texto del segundo parámetro.
 *   - Si el número es múltiplo de 3 y de 5, muestra las dos cadenas de texto concatenadas.
 *   - La función retorna el número de veces que se ha impreso el número en lugar de los textos.
 *
 * Presta especial atención a la sintaxis que debes utilizar en cada uno de los casos.
 * Cada lenguaje sigue una convenciones que debes de respetar para que el código se entienda.
 */

//Funcion tradicional
function tradicional() {
  console.log("Función creada de manera tradicional");
}

tradicional();

//funciones de flecha

const sinParamsNiReturns = () => {
  console.log("Esta función no tiene parametros ni retornos");
};

sinParamsNiReturns();

const conParametros = (num1, num2) => {
  console.log(num1 + num2);
};

conParametros(27, 7);

const conRetorno = (num1, num2) => {
  return num1 + num2;
};

conRetorno(9, 8);

//Funciones dentro de funciones?

const miPrimerFuncion = (nombre, apellido, edad) => {
  const persona = [nombre, apellido, edad];

  const miSegundaFuncion = (persona) => {
    console.log(persona);
    return (
      "Mi nombre es : " +
      persona[0] +
      ", mi apellido es: " +
      persona[1] +
      " y mi edad es: " +
      persona[2]
    );
  };

  return miSegundaFuncion(persona);
};

miPrimerFuncion("Andrew", "Codev", 31);

//Variable local y global
let varGlobal = "Soy global";

const localGlobal = (varGlobal) => {
  let varLocal = "Soy Local";
  if (varGlobal === varLocal) {
    return "La variable local y la global son iguales";
  } else {
    return "La variable local y la global son diferentes";
  }
};

localGlobal(varGlobal);

// DIFICULTAD EXTRA
const dificultadExtra = (cad1, cad2) => {
  let contador = 0;

  for (let i = 0; i <= 100; i++) {
    const esMultiploDe3 = i % 3 === 0;
    const esMultiploDe5 = i % 5 === 0;

    if (esMultiploDe3 && esMultiploDe5) {
      console.log(`${cad1} ${cad2}`);
    } else if (esMultiploDe3) {
      console.log(`${cad1}`);
    } else if (esMultiploDe5) {
      console.log(`${cad2}`);
    } else {
      console.log(i);
      contador++;
    }
  }

  return contador;
};

const vecesNumero = dificultadExtra("Hola", "Mundo");
console.log(`Números impresos: ${vecesNumero}`);
